Add an option in the astro config build object to set concurrency

jwickers opened the pull request:
## Changes

This makes concurrency optional using the implementation that was removed before.
By default still use a concurrency of 1 and does not change anything in the build.

One can set it in their `astro.config.mjs`:
```
export default defineConfig({
	// integrations: [...],
	build: {
		concurrency: 8,
	}
	// adapter: ...
});
```

Signed-off-by: Jeremy Wickersheimer <jwickers@gmail.com>

## Testing

Tested with a very large build that has >300k pages (mostly all in one URL pattern).
Manages to reduce the total build from 6000s to 3000s with a concurrency of 8.

## Docs

May need a Doc update to mention the new setting.

matthewp reviewed (commented):
Awesome! We used to do this by default, I'm not sure why that changed.

matthewp commented on .changeset/healthy-cougars-relax.md:
```diff
@@ -0,0 +1,5 @@
+'astro': minor
+---
+
+Add build concurrency config option
```
The `buildConfig`, which is where you are setting this, is not part of the config object. It's something that can be modified via an integration. For example you could do this:

```js
export default defineConfig({
  integrations: [
    {
      name: 'my-integration',
      hooks: {
        'astro:build:start': ({ buildConfig }) {
          buildConfig.concurrency = 8;
        }
      }
    }
  ]
});
```

We've never discussed making the `buildConfig` be exposed by the top-level config object. I think that might require an RFC if it led to nesting, which we don't currently have. I think leaving this as an advanced API that you can add via an integration is a better way to go, for now.

jwickers replied:
Sure, can be changed to use `opts.astroConfig` and move `concurrency` from `build` to the root of `AstroConfigSchema` then?

In that case, it would be defined in:
```
export default defineConfig({
	// integrations: [...],
	concurrency: 8,
	// adapter: ...
});
```

We could also rename the key in that case if you prefer.

matthewp replied:
We can move it to the top-level config, but that's going to take some convincing with the core team. I'll bring it up to them.

FredKSchott replied:
I'm a light +1 on a top-level `build.concurrency` astro config value, and definitely prefer it to the hidden integration-only `buildConfig` object

matthewp reviewed (changes requested):
The API as described in the PR is not how this actually works. Need to figure out what the API should be.

jwickers commented:
I've also changed the way tasks are processed to use Promise.race instead of Promise.all as this allows the tasks to be queued as soon as one completes instead of batching them in a fixed array where Promise.all would 'block' until the batch was completed.

natemoo-re reviewed (approved):
This all LGTM now! I'll defer to @matthewp!

matthewp commented:
Let's move the config option to `build.concurrency` as @FredKSchott suggested, then we can merge. cc @jwickers

natemoo-re commented:
@matthewp from my understanding that was already done? https://github.com/withastro/astro/pull/4052/files#diff-10cec9990a82f35ed4ce5e0137005fc5035b0291910a4162f1ff0c5adbf2a8efR171

matthewp commented:
It's been 2 weeks without an update so I'm going to close this PR. Please feel free to reopen if you change where the config option lives. Thanks!

jwickers commented:
Hi, sorry, I've been out for the summer. I rebased it to the latest main branch and checked that it was still working for my use case.

I've updated the PR description to where the config lives, I had moved it to be in the main config object already. But let me know if you meant to move it somewhere else.
